File: web/state.ts
import type { Quality, ScanState } from "./types";
export class CaptureState {
  value: ScanState = {
    type: "state",
    phase: "red",
    message: "Camera ready. Choose Start scanning on your computer.",
    paused: true,
    activeId: null,
    lastSaved: null,
    lastCapture: null,
    retakeOf: null,
    armed: true,
    cameraConnected: true,
    streamFresh: true,
    detectorReady: false,
    count: 0,
    quality: {
      ok: false,
      quad: null,
      hands: [],
      reason: "Loading image checks.",
    },
  };
  private stable = 0;
  private frames = 0;
  private absent = 0;
  private feedback: {
    phase: "red" | "amber";
    message: string;
    since: number;
  } | null = null;
  private latched = false;
  private reset() {
    this.stable = 0;
    this.frames = 0;
  }
  private showFeedback(phase: "red" | "amber", message: string, now: number) {
    if (this.feedback?.phase !== phase || this.feedback.message !== message)
      this.feedback = { phase, message, since: now };
    if (now - this.feedback.since >= 250) {
      this.value.phase = phase;
      this.value.message = message;
    }
  }
  control(action: string) {
    this.feedback = null;
    if (this.value.activeId) return;
    if (
      (action === "start" || action === "retry") &&
      this.value.recovery === "upload"
    )
      return;
    if (
      action === "retry" &&
      !this.value.lastCapture &&
      this.value.recovery !== "retake"
    )
      return;
    this.reset();
    if (action === "pause") {
      this.value.paused = true;
      this.value.phase = "red";
      this.value.message = "Paused.";
    }
    if (action === "start" || action === "retry") {
      this.value.needsAttention = false;
      this.value.recovery = undefined;
      this.value.paused = false;
      this.latched = false;
      if (action === "retry") {
        this.value.retakeOf = this.value.lastCapture;
        this.value.armed = true;
      }
      this.value.phase = "red";
      this.value.message = this.value.retakeOf
        ? "Retaking the same receipt. Keep it in view and hold steady."
        : "Place one receipt on the dark background.";
    }
  }
  observe(q: Quality, now: number): string | null {
    this.value.quality = q;
    if (this.value.activeId) return null;
    // Removal also clears retake intent while paused or waiting after a failed check.
    // A later receipt must never inherit the previous receipt's identity.
    if (this.value.lastCapture || !this.value.armed) {
      if (q.empty && !q.hands.length) {
        this.absent ||= now;
        if (now - this.absent >= 450) {
          this.value.lastCapture = null;
          this.value.retakeOf = null;
          this.value.armed = true;
          this.reset();
          if (!this.latched && !this.value.paused) {
            this.value.phase = "red";
            this.value.message = "Ready for the next receipt.";
          }
        }
      } else this.absent = 0;
    }
    if (this.value.paused || this.latched || !this.value.armed) return null;
    if (q.empty && !q.hands.length) {
      this.reset();
      this.feedback = null;
      this.value.phase = "red";
      this.value.message = this.value.lastSaved
        ? "Ready for the next receipt."
        : q.reason;
      return null;
    }
    if (!q.ok || !q.quad || (q.motion ?? 0) > 3.5) {
      this.reset();
      this.showFeedback("red", q.reason, now);
      return null;
    }
    // The worker compares aligned paper content. Handheld translation must not
    // reset stability merely because the same sharp paper moved in the frame.
    if (this.frames === 0) this.stable = now;
    this.frames++;
    this.showFeedback("amber", "Checking image stability…", now);
    if (now - this.stable >= 900 && this.frames >= 4) {
      this.value.phase = "amber";
      this.value.activeId = crypto.randomUUID();
      this.value.message = "Taking the photo…";
      return this.value.activeId;
    }
    return null;
  }
  saved(id: string, count?: number) {
    this.feedback = null;
    if (count !== undefined) this.value.count = count;
    else if (this.value.lastSaved !== id) this.value.count++;
    this.value.stage = undefined;
    this.value.needsAttention = false;
    this.value.activeId = null;
    this.value.lastSaved = id;
    this.value.lastCapture = id;
    this.value.retakeOf = null;
    this.value.recovery = undefined;
    this.value.armed = false;
    this.absent = 0;
    this.latched = false;
    this.value.phase = "green";
    this.value.message = "Saved privately. Remove the receipt, then next.";
    this.reset();
  }
  failed(message: string, recovery?: "retake" | "upload", retainedId?: string) {
    if (retainedId) {
      this.value.lastCapture = retainedId;
      this.value.retakeOf = null;
      this.value.armed = false;
      this.absent = 0;
    }
    this.value.recovery = recovery;
    this.feedback = null;
    this.value.stage = undefined;
    this.value.needsAttention = true;
    this.value.activeId = null;
    this.latched = true;
    this.value.phase = "red";
    this.value.message = message;
    this.reset();
  }

  interrupt() {
    // Re-establish stability and continuous paper removal after a connection gap.
    // Keep saved/failed captures and the removal latch intact.
    this.absent = 0;
    this.reset();
  }
}

File: server/state.py
"""Pure capture state machine. Only persisted captures may enter green."""

from dataclasses import dataclass, field
from uuid import uuid4


@dataclass
class CaptureState:
    paused: bool = True
    phase: str = "red"
    message: str = "Connect the phone, then start scanning."
    active_id: str | None = None
    armed: bool = True
    stable_since: float | None = None
    absent_since: float | None = None
    stable_frames: int = 0
    last_quad: list[list[float]] | None = None
    last_seen: float = 0
    capture_started: float = 0
    last_saved: str | None = None
    error_latched: bool = False
    quality: dict = field(default_factory=dict)

    def reset_stability(self) -> None:
        self.stable_since = None
        self.stable_frames = 0
        self.last_quad = None

    def control(self, action: str) -> None:
        if self.active_id:
            raise ValueError("Wait for the current capture to finish.")
        if action == "pause":
            self.paused = True
            self.phase, self.message = "red", "Paused."
        elif action in {"start", "retry"}:
            self.paused = False
            self.error_latched = False
            if action == "retry":
                self.armed = True
            self.phase, self.message = "red", "Place one receipt on the dark background."
        else:
            raise ValueError("Unknown control.")
        self.reset_stability()

    def observe(self, detection: dict, now: float) -> str | None:
        self.last_seen = now
        self.quality = detection
        if self.active_id or self.paused or self.error_latched:
            return None
        # A hand/noisy boundary is not removal. Require a clearly empty, still desk.
        if not self.armed:
            self.phase = "green"
            self.message = "Saved on this computer. Remove the receipt, then next."
            if detection.get("empty"):
                if self.absent_since is None:
                    self.absent_since = now
                if now - self.absent_since >= 0.45:
                    self.armed = True
                    self.reset_stability()
                    self.phase, self.message = "red", "Ready for the next receipt."
            else:
                self.absent_since = None
            return None
        if not detection.get("ok"):
            self.reset_stability()
            self.phase, self.message = "red", detection["reason"]
            return None
        quad = detection["quad"]
        changed = (
            self.last_quad is None
            or max(
                abs(a - b)
                for previous, current in zip(self.last_quad, quad, strict=True)
                for a, b in zip(previous, current, strict=True)
            )
            > 0.012
        )
        if changed or detection.get("motion", 0) > 3.5:
            self.stable_since, self.stable_frames = now, 0
        self.last_quad = quad
        self.stable_frames += 1
        self.phase, self.message = "amber", "Hold still—checking the receipt."
        if self.stable_since is None:
            self.stable_since = now
        if now - self.stable_since >= 0.9 and self.stable_frames >= 4:
            self.active_id = str(uuid4())
            self.capture_started = now
            self.message = "Capturing and saving—do not move the receipt."
            return self.active_id
        return None

    def saved(self, capture_id: str) -> None:
        if capture_id != self.active_id:
            raise ValueError("Capture is no longer active.")
        self.last_saved = capture_id
        self.error_latched = False
        self.active_id = None
        self.armed = False
        self.absent_since = None
        self.phase, self.message = "green", "Saved on this computer. Remove the receipt, then next."
        self.reset_stability()

    def failed(self, message: str) -> None:
        self.active_id = None
        self.phase, self.message = "red", message
        self.error_latched = True
        self.reset_stability()

    def disconnected(self) -> None:
        self.reset_stability()
        self.phase, self.message = "red", "Phone disconnected. Reconnect before moving on."

    def public(self) -> dict:
        return {
            "phase": self.phase,
            "message": self.message,
            "paused": self.paused,
            "activeId": self.active_id,
            "lastSaved": self.last_saved,
            "armed": self.armed,
            "quality": self.quality,
        }
